### ISSUE TITLE
Unable to (cross-/)compile OpenPose for TK1

### ISSUE THREAD

**Issue by alvaroga91** [CMU-Perceptual-Computing-Lab/openpose#478]
### Issue Summary
Hi there, 
I've been trying to have OpenPose working in the TK1, CUDA 6.5, but after several attemps I've been able to isolate the main issue and I come here seeking for some help. Needless to say, I've already read _several times_ all the issues regarding the TK1 (mainly [this](https://github.com/CMU-Perceptual-Computing-Lab/openpose/issues/44) and [this](https://github.com/CMU-Perceptual-Computing-Lab/openpose/issues/58)).

In both of the following situations, I've managed to reproduce the error by trying to compile a single .cu (openpose/src/openpose/core/resizeAndMergeBase.cu, but tested with others with same outcome) from OpenPose through command line (with a nvcc command and its corresponding flags/paths). This is what I'll be posting, but take into account that the same error appears when I launch "make all" in my OpenPose build, but this is much more readable imho

Also, I've been able to compile natively in both a VM Ubuntu 14.04_x86 and "Real machine" 16.04_x64, so I have some knowledge of the OpenPose environment. In both scenarios, compilation went flawlessly but in the VM it never launched since VMs do not offer CUDA support. 

For this process of trying to compile  I've followed 2 ways with different errors: Native compilation and cross-compilation.

### Native complation testing

Native compilation in the TK1. I could install all the dependencies with apt-get and got Caffe compiled as well. But when nvcc in OpenPose comes to play... error. 

This was the command tested

$ /usr/local/cuda-6.5/bin/nvcc -I/home/ubuntu/OpenPose/openpose/include -std=c++11 -G -g -O0 --compile --relocatable-device-code=false -gencode arch=compute_20,code=compute_20 -gencode arch=compute_30,code=sm_30 -gencode arch=compute_50,code=sm_50 -gencode arch=compute_52,code=sm_52 -x cu -o  Test.o resizeAndMergeBase.cu 

And the error was: 
resizeAndMergeBase.cu(80): internal error: assertion failed at: "/dvs/p4/build/sw/rel/gpu_drv/r343/r343_00/drivers/compiler/edg/EDG_4.8/src/decl_inits.c", line 3135

Unluckily, I was not only the one with these error overall in the internet with pretty much no solution:
[Post 1](https://devtalk.nvidia.com/default/topic/988913/jetson-tk1/a-catastrophic-error-detected-in-the-compilation-of-cuckoohash-/)
[Post 2](http://blog.csdn.net/shenck1992/article/details/59113298)
[Post 3](https://github.com/CMU-Perceptual-Computing-Lab/openpose/issues/44)
[Post 4 ](https://github.com/tensorflow/tensorflow/issues/3786)

This is what I got with **-DCUDA_VERBOSE_BUILD=ON.**

ubuntu@tegra-ubuntu:~/OpenPose/openpose/build$ make all
/usr/bin/cmake -H/home/ubuntu/OpenPose/openpose -B/home/ubuntu/OpenPose/openpose/build --check-build-system CMakeFiles/Makefile.cmake 0
/usr/bin/cmake -E cmake_progress_start /home/ubuntu/OpenPose/openpose/build/CMakeFiles /home/ubuntu/OpenPose/openpose/build/CMakeFiles/progress.marks
make -f CMakeFiles/Makefile2 all
make[1]: Entering directory `/home/ubuntu/OpenPose/openpose/build'
make -f src/openpose/CMakeFiles/openpose.dir/build.make src/openpose/CMakeFiles/openpose.dir/depend
make[2]: Entering directory `/home/ubuntu/OpenPose/openpose/build'
/usr/bin/cmake -E cmake_progress_report /home/ubuntu/OpenPose/openpose/build/CMakeFiles 8
[  1%] Building NVCC (Device) object src/openpose/CMakeFiles/openpose.dir/hand/./openpose_generated_renderHand.cu.o
cd /home/ubuntu/OpenPose/openpose/build/src/openpose/CMakeFiles/openpose.dir/hand && /usr/bin/cmake -E make_directory /home/ubuntu/OpenPose/openpose/build/src/openpose/CMakeFiles/openpose.dir/hand/.
cd /home/ubuntu/OpenPose/openpose/build/src/openpose/CMakeFiles/openpose.dir/hand && /usr/bin/cmake -D verbose:BOOL=ON -D build_configuration:STRING= -D generated_file:STRING=/home/ubuntu/OpenPose/openpose/build/src/openpose/CMakeFiles/openpose.dir/hand/./openpose_generated_renderHand.cu.o -D generated_cubin_file:STRING=/home/ubuntu/OpenPose/openpose/build/src/openpose/CMakeFiles/openpose.dir/hand/./openpose_generated_renderHand.cu.o.cubin.txt -P /home/ubuntu/OpenPose/openpose/build/src/openpose/CMakeFiles/openpose.dir/hand/openpose_generated_renderHand.cu.o.cmake
-- Removing /home/ubuntu/OpenPose/openpose/build/src/openpose/CMakeFiles/openpose.dir/hand/./openpose_generated_renderHand.cu.o
/usr/bin/cmake -E remove /home/ubuntu/OpenPose/openpose/build/src/openpose/CMakeFiles/openpose.dir/hand/./openpose_generated_renderHand.cu.o
-- Generating dependency file: /home/ubuntu/OpenPose/openpose/build/src/openpose/CMakeFiles/openpose.dir/hand/openpose_generated_renderHand.cu.o.NVCC-depend
/usr/local/cuda-6.5/bin/nvcc -M -D__CUDACC__ /home/ubuntu/OpenPose/openpose/src/openpose/hand/renderHand.cu -o /home/ubuntu/OpenPose/openpose/build/src/openpose/CMakeFiles/openpose.dir/hand/openpose_generated_renderHand.cu.o.NVCC-depend -ccbin /usr/bin/cc -m32 -DUSE_CAFFE -DUSE_CUDA -Dopenpose_EXPORTS -Xcompiler ,\"-fPIC\",\"-g\" -gencode arch=compute_32,code=sm_32 -Xcudafe --diag_suppress=cc_clobber_ignored -Xcudafe --diag_suppress=integer_sign_change -Xcudafe --diag_suppress=useless_using_declaration -Xcudafe --diag_suppress=set_but_not_used -std=c++11 -DNVCC -I/usr/local/cuda-6.5/include -I/usr/include/opencv -I/usr/include -I/home/ubuntu/OpenPose/openpose/include -I/home/ubuntu/OpenPose/openpose/”/home/ubuntu/OpenPose/openpose/3rdparty/caffe/include” -I/home/ubuntu/OpenPose/openpose/PUBLIC -I/usr/local/cuda-6.5/include
-- Generating temporary cmake readable file: /home/ubuntu/OpenPose/openpose/build/src/openpose/CMakeFiles/openpose.dir/hand/openpose_generated_renderHand.cu.o.depend.tmp
/usr/bin/cmake -D input_file:FILEPATH=/home/ubuntu/OpenPose/openpose/build/src/openpose/CMakeFiles/openpose.dir/hand/openpose_generated_renderHand.cu.o.NVCC-depend -D output_file:FILEPATH=/home/ubuntu/OpenPose/openpose/build/src/openpose/CMakeFiles/openpose.dir/hand/openpose_generated_renderHand.cu.o.depend.tmp -P /usr/share/cmake-2.8/Modules/FindCUDA/make2cmake.cmake
-- Copy if different /home/ubuntu/OpenPose/openpose/build/src/openpose/CMakeFiles/openpose.dir/hand/openpose_generated_renderHand.cu.o.depend.tmp to /home/ubuntu/OpenPose/openpose/build/src/openpose/CMakeFiles/openpose.dir/hand/openpose_generated_renderHand.cu.o.depend
/usr/bin/cmake -E copy_if_different /home/ubuntu/OpenPose/openpose/build/src/openpose/CMakeFiles/openpose.dir/hand/openpose_generated_renderHand.cu.o.depend.tmp /home/ubuntu/OpenPose/openpose/build/src/openpose/CMakeFiles/openpose.dir/hand/openpose_generated_renderHand.cu.o.depend
-- Removing /home/ubuntu/OpenPose/openpose/build/src/openpose/CMakeFiles/openpose.dir/hand/openpose_generated_renderHand.cu.o.depend.tmp and /home/ubuntu/OpenPose/openpose/build/src/openpose/CMakeFiles/openpose.dir/hand/openpose_generated_renderHand.cu.o.NVCC-depend
/usr/bin/cmake -E remove /home/ubuntu/OpenPose/openpose/build/src/openpose/CMakeFiles/openpose.dir/hand/openpose_generated_renderHand.cu.o.depend.tmp /home/ubuntu/OpenPose/openpose/build/src/openpose/CMakeFiles/openpose.dir/hand/openpose_generated_renderHand.cu.o.NVCC-depend
-- Generating /home/ubuntu/OpenPose/openpose/build/src/openpose/CMakeFiles/openpose.dir/hand/./openpose_generated_renderHand.cu.o
/usr/local/cuda-6.5/bin/nvcc /home/ubuntu/OpenPose/openpose/src/openpose/hand/renderHand.cu -c -o /home/ubuntu/OpenPose/openpose/build/src/openpose/CMakeFiles/openpose.dir/hand/./openpose_generated_renderHand.cu.o -ccbin /usr/bin/cc -m32 -DUSE_CAFFE -DUSE_CUDA -Dopenpose_EXPORTS -Xcompiler ,\"-fPIC\",\"-g\" -gencode arch=compute_32,code=sm_32 -Xcudafe --diag_suppress=cc_clobber_ignored -Xcudafe --diag_suppress=integer_sign_change -Xcudafe --diag_suppress=useless_using_declaration -Xcudafe --diag_suppress=set_but_not_used -std=c++11 -DNVCC -I/usr/local/cuda-6.5/include -I/usr/include/opencv -I/usr/include -I/home/ubuntu/OpenPose/openpose/include -I/home/ubuntu/OpenPose/openpose/”/home/ubuntu/OpenPose/openpose/3rdparty/caffe/include” -I/home/ubuntu/OpenPose/openpose/PUBLIC -I/usr/local/cuda-6.5/include
/home/ubuntu/OpenPose/openpose/include/openpose/hand/handParameters.hpp(15): internal error: assertion failed at: "/dvs/p4/build/sw/rel/gpu_drv/r343/r343_00/drivers/compiler/edg/EDG_4.8/src/decl_inits.c", line 3135

1 catastrophic error detected in the compilation of "/tmp/tmpxft_00001206_00000000-6_renderHand.cpp1.ii".
Compilation aborted.
Aborted
-- Removing /home/ubuntu/OpenPose/openpose/build/src/openpose/CMakeFiles/openpose.dir/hand/./openpose_generated_renderHand.cu.o
/usr/bin/cmake -E remove /home/ubuntu/OpenPose/openpose/build/src/openpose/CMakeFiles/openpose.dir/hand/./openpose_generated_renderHand.cu.o
CMake Error at openpose_generated_renderHand.cu.o.cmake:264 (message):
  Error generating file
  /home/ubuntu/OpenPose/openpose/build/src/openpose/CMakeFiles/openpose.dir/hand/./openpose_generated_renderHand.cu.o

make[2]: *** [src/openpose/CMakeFiles/openpose.dir/hand/./openpose_generated_renderHand.cu.o] Error 1
make[2]: Leaving directory `/home/ubuntu/OpenPose/openpose/build'
make[1]: *** [src/openpose/CMakeFiles/openpose.dir/all] Error 2
make[1]: Leaving directory `/home/ubuntu/OpenPose/openpose/build'
make: *** [all] Error 2

**-------OUTPUT ENDS HERE**

Even though there were some comments from users [airobots](https://github.com/CMU-Perceptual-Computing-Lab/openpose/issues/44#issuecomment-329639845) and [cortinas](https://github.com/CMU-Perceptual-Computing-Lab/openpose/issues/58#issuecomment-307590244) (what a troll, arch 53 does not even work with CUDA 6.5, and the higher version of CUDNN you can install in TK1 is v2, which does not work with OpenPose), I was unable to have it working.

Nevertheless, I decided to take the harder and longer route: Cross-compilation. 

### Cross compilation testing

I must say that crosscompiling all the requisites for Caffe have been a pain in the *** which involved lots of git cloning, library de-referencing and referencing, makefiles editing, scp'ing the libraries and headers from the TK1 to take the shorter way... At the end?The dependencies were slowly but steady solved, but at the end I got a bit fed up and never got Caffe compiled there, just compiled it in the TK1, tar'd the Caffe folder and brought it back. 

Back to CMake, lots of edition in order to change the binaries and libraries used from the native x86 to ARM cross-compiling (gcc-linaro 4.8.5). I finally managed no complains from CMake, and could generate the Makefile. But.. another error appeared. I managed to reproduce it again, but with this command: 

/usr/local/cuda-6.5/bin/nvcc -ccbin /home/metal/Downloads/gcc-linaro-4.8-2015.06-x86_64_arm-linux-gnueabihf/bin/arm-linux-gnueabihf-gcc -I/home/metal/OpenPose/arm/libs/include -I/home/metal/OpenPose/openpose/include  -G -g -O0 --compile --relocatable-device-code=false -std=c++11 -gencode arch=compute_20,code=compute_20 -gencode arch=compute_20,code=sm_20 -x cu -o Test.o resizeAndMergeBase.cu

And the outcome was: 

/home/metal/Downloads/gcc-linaro-4.8-2015.06-x86_64_arm-linux-gnueabihf/arm-linux-gnueabihf/include/c++/4.8.5/new(91): error: first parameter of allocation function must be of type "size_t"

/home/metal/Downloads/gcc-linaro-4.8-2015.06-x86_64_arm-linux-gnueabihf/arm-linux-gnueabihf/include/c++/4.8.5/new(93): error: first parameter of allocation function must be of type "size_t"

/home/metal/Downloads/gcc-linaro-4.8-2015.06-x86_64_arm-linux-gnueabihf/arm-linux-gnueabihf/include/c++/4.8.5/new(99): error: first parameter of allocation function must be of type "size_t"

/home/metal/Downloads/gcc-linaro-4.8-2015.06-x86_64_arm-linux-gnueabihf/arm-linux-gnueabihf/include/c++/4.8.5/new(101): error: first parameter of allocation function must be of type "size_t"

/home/metal/Downloads/gcc-linaro-4.8-2015.06-x86_64_arm-linux-gnueabihf/arm-linux-gnueabihf/include/c++/4.8.5/new(109): error: first parameter of allocation function must be of type "size_t"

/home/metal/Downloads/gcc-linaro-4.8-2015.06-x86_64_arm-linux-gnueabihf/arm-linux-gnueabihf/include/c++/4.8.5/new(111): error: first parameter of allocation function must be of type "size_t"

/home/metal/OpenPose/arm/libs/include/common_functions.h(91): error: first parameter of allocation function must be of type "size_t"

/home/metal/OpenPose/arm/libs/include/common_functions.h(92): error: first parameter of allocation function must be of type "size_t"

/home/metal/OpenPose/arm/libs/include/common_functions.h(145): error: first parameter of allocation function must be of type "size_t"

/home/metal/OpenPose/arm/libs/include/common_functions.h(146): error: first parameter of allocation function must be of type "size_t"

10 errors detected in the compilation of "/tmp/tmpxft_00008b7a_00000000-6_resizeAndMergeBase.cpp1.ii".

**HOWEVER** I relaunched the command with no --ccbin (native gcc 4.8.4)

/usr/local/cuda-6.5/bin/nvcc -I/home/metal/OpenPose/arm/libs/include -I/home/metal/OpenPose/openpose/include  -G -g -O0 --compile --relocatable-device-code=false -std=c++11 -gencode arch=compute_20,code=compute_20 -gencode arch=compute_20,code=sm_20 -x cu -o Test.o resizeAndMergeBase.cu
(I left the arm headers on purpose)

**AND IT WORKED**. Just some warnings (warning: integer conversion resulted in a change of sign) but compilation worked fine and my Test.o was created. 

I even tested with a more recent gcc-linaro 4.9.4 (which led me to hack the CUDA host_config header file to remove the error), which gave me quite a few errors more apart from the ones found in the 4.8.5:

/home/metal/Downloads/gcc-linaro-4.9-2016.02-x86_64_arm-linux-gnueabihf/arm-linux-gnueabihf/include/c++/4.9.4/bits/alloc_traits.h(248): error: expected a ">"
          detected during instantiation of type "std::allocator_traits<_Alloc>::__has_construct<_Tp, _Args...>" 
(251): here

/home/metal/Downloads/gcc-linaro-4.9-2016.02-x86_64_arm-linux-gnueabihf/arm-linux-gnueabihf/include/c++/4.9.4/type_traits(130): error: not a class or struct name
          detected during:
            instantiation of class "std::__and_<_B1> [with _B1=<error-type>]" 
(1957): here
            instantiation of type "std::_Require<<error-type>>" 
/home/metal/Downloads/gcc-linaro-4.9-2016.02-x86_64_arm-linux-gnueabihf/arm-linux-gnueabihf/include/c++/4.9.4/bits/alloc_traits.h(251): here

/home/metal/Downloads/gcc-linaro-4.9-2016.02-x86_64_arm-linux-gnueabihf/arm-linux-gnueabihf/include/c++/4.9.4/type_traits(1957): error: class "std::__and_<<error-type>>" has no member "value"
          detected during instantiation of type "std::_Require<<error-type>>" 
/home/metal/Downloads/gcc-linaro-4.9-2016.02-x86_64_arm-linux-gnueabihf/arm-linux-gnueabihf/include/c++/4.9.4/bits/alloc_traits.h(251): here

/home/metal/Downloads/gcc-linaro-4.9-2016.02-x86_64_arm-linux-gnueabihf/arm-linux-gnueabihf/include/c++/4.9.4/type_traits(1957): error: class "std::enable_if<<error-constant>, void>" has no member "type"
          detected during instantiation of type "std::_Require<<error-type>>" 
/home/metal/Downloads/gcc-linaro-4.9-2016.02-x86_64_arm-linux-gnueabihf/arm-linux-gnueabihf/include/c++/4.9.4/bits/alloc_traits.h(251): here

14 errors detected in the compilation of "/tmp/tmpxft_00008d23_00000000-6_resizeAndMergeBase.cpp1.ii".

It has been a very long week sorting and solving many walls (you know, another day in the office), but I feel that the solution is right there, and it is the difference between having OpenPose working in the TK1 or not :') 

Thanks

**Comment by alvaroga91** [comment]
I'm betting that it might have something to do with this: 

https://devtalk.nvidia.com/default/topic/865183/jetson-tk1-problem-compiling-with-nvcc-when-including-lt-future-gt-/?offset=7

Although instead of "future", it will be "new".

**Comment by alvaroga91** [comment]
I don't wether if I should cry or laugh... I finally managed it. A friend of mine mentioned "Why don't you just update to Cuda 7?", which answer is "due to the 32bit limitation of the TK1", which led me to think to try the above command with -m32 in my host machine.

And it worked (No luck in the native TK1 compiler tho)

After that I kept playing with the CMake-gui, in order to find the good set of flags:

I had to turn off CUDA_64_BIT_DEVICE_CODE, which automatically adds the -m32 flag (Duh!). No need to add the flag manually elsewhere.

However I was having a problem with the -ccbin flag of NVCC. Also, I had to leave the CUDA_HOST_COMPILATION_CPP on, with no path in the CUDA_HOST_COMPILER in order to CMake no complain. But still, the flag -ccbin kept appearing, prompting errors using what was following the flag as the host compiler (Making the compilation fail).

After some research, I could see that I was not the only one and it's a bug from CMake (or maybe it was my fault, at this point I don't really care...). I modified the CMake files in order to not output ever the -cbin.

In CMake 2.8, these are files and the lines to comment:

/usr/share/cmake-2.8/Modules/FindCUDA/run_nvcc,cmake, lines 111 and 113
/usr/share/cmake-2.8/Modules/FindCUDA.cmake, line 1377

With this, the -ccbin didn't appear and the compilation advanced!!! (Still managing other errors, now in the protobuf. An old friend of mine but I have to recompile Caffe again with the updated version from the TK1).

I'll update with any progress.

**Comment by alvaroga91** [comment]
Well, made quite some progress! Now stuck when building target openpose and linking the bins. 

Many undefined references, probably because I've used a different compiler for Caffe. I'll try to Compile Caffe in the same machine. However if someone could confirm me this, it would be pretty nice! (I've also brought all the libs from the TK1, could also that be that problem? I had these before moving the libs, I would say even more) 
[failOpenPose.txt](https://github.com/CMU-Perceptual-Computing-Lab/openpose/files/1800584/failOpenPose.txt)

**Comment by alvaroga91** [comment]
Long story short: I got it compiled through cross-compilation. A real pain but it worked. However, CUDNN is not supported (In TK1 is v2), so if you manage to get it compiled (without CUDNN) good luck trying to execute a 10GB~ software in a 2GB memory. I didn't realised this until recently (I even try killing X and all). Until CUDNNv2 is supported, OpenPose in a TK1 is a no-no.

TL;DR: Cross-compiled succesfully but since CUDNNv2 is not supported, no way it works in the 2GB RAM TK1.

**Comment by gineshidalgo99** [comment]
No idea about TK1, but if cuDNN is not available, your second option is running with very small net resolution, something like `net_resolution -1x80`, -1x160, etc. It will still work for very close people (it will fail for small people)
